kernel/src/predicates/mod.rs

```diff
@@ -0,0 +1,652 @@
+use crate::expressions::{
+    BinaryOperator, ColumnName, Expression as Expr, Scalar, UnaryOperator, VariadicOperator,
+};
+use crate::schema::DataType;
+
+use std::cmp::Ordering;
+use tracing::debug;
+
+#[cfg(test)]
+mod tests;
+
+/// Evaluates a predicate expression tree against column names that resolve as scalars. Useful for
+/// testing/debugging but also serves as a reference implementation that documents the expression
+/// semantics that kernel relies on for data skipping.
+///
+/// # Inverted expression semantics
+///
+/// Because inversion(`NOT` operator) has special semantics and can often be optimized away by
+/// pushing it down, most methods take an `inverted` flag. That allows operations like
+/// [`UnaryOperator::Not`] to simply evaluate their operand with a flipped `inverted` flag,
+///
+/// # NULL and error semantics
+///
+/// Literal NULL values almost always produce cascading changes in the predicate's structure, so we
+/// represent them by `Option::None` rather than `Scalar::Null`. This allows e.g. `A < NULL` to be
+/// rewritten as `NULL`, or `AND(NULL, FALSE)` to be rewritten as `FALSE`.
```

Comment:
(even if something does produce `Scalar::Null` tho, it should only risk giving up some of those "structural" plan simplifications. It would still evaluate to a correct result)